src/twisted/internet/defer.py

```diff
@@ -515,54 +542,151 @@ def addCallbacks(
-        ],
-        *args: object,
-        **kwargs: object,
+        errback: Callable[Concatenate[Failure, _P], Deferred[_NextDeferredResultT]],
+        *args: _P.args,
+        **kwargs: _P.kwargs,
+    ) -> "Deferred[Union[_DeferredResultT, _NextDeferredResultT]]":
+        ...
+
+    @overload
+    def addErrback(
+        self,
+        errback: Callable[Concatenate[Failure, _P], Failure],
+        *args: _P.args,
+        **kwargs: _P.kwargs,
+    ) -> "Deferred[Union[_DeferredResultT]]":
+        ...
+
+    @overload
+    def addErrback(
+        self,
+        errback: Callable[Concatenate[Failure, _P], _NextDeferredResultT],
+        *args: _P.args,
+        **kwargs: _P.kwargs,
+    ) -> "Deferred[Union[_DeferredResultT, _NextDeferredResultT]]":
+        ...
+
+    def addErrback(
+        self,
+        errback: Callable[..., Any],
+        *args: Any,
+        **kwargs: Any,
     ) -> "Deferred[Union[_DeferredResultT, _NextDeferredResultT]]":
         """
         Convenience method for adding just an errback.
 
         See L{addCallbacks}.
         """
-        # type note: passthru constrains the type of errback in a way which mypy
-        #     can't propagate through to _NextDeferredResultT, so we have to
-        #     ignore a type error.
         return self.addCallbacks(
             passthru,
-            errback,  # type: ignore[arg-type]
+            errback,
             errbackArgs=args,
             errbackKeywords=kwargs,
         )
 
+    @overload
+    def addBoth(
+        self,
+        callback: Callable[Concatenate[Union[_DeferredResultT, Failure], _P], Failure],
+        *args: _P.args,
+        **kwargs: _P.kwargs,
+    ) -> Deferred[_NextDeferredResultT]:
+        ...
+
+    @overload
+    def addBoth(
+        self,
+        callback: Callable[
+            Concatenate[Union[_DeferredResultT, Failure], _P],
+            Deferred[_NextDeferredResultT],
+        ],
+        *args: _P.args,
+        **kwargs: _P.kwargs,
+    ) -> Deferred[_NextDeferredResultT]:
+        ...
+
+    @overload
     def addBoth(
         self,
         callback: Callable[
-            ...,
-            "Union[_NextDeferredResultT, Deferred[_NextDeferredResultT]]",
+            Concatenate[Union[_DeferredResultT, Failure], _P],
+            Union[_DeferredResultT, Failure],
         ],
-        *args: object,
-        **kwargs: object,
+        *args: _P.args,
+        **kwargs: _P.kwargs,
+    ) -> Deferred[_NextDeferredResultT]:
+        ...
+
+    @overload
+    def addBoth(
+        self,
+        callback: Callable[
+            Concatenate[Union[_DeferredResultT, Failure], _P], _NextDeferredResultT
+        ],
+        *args: _P.args,
+        **kwargs: _P.kwargs,
+    ) -> Deferred[_NextDeferredResultT]:
+        ...
+
+    @overload
+    def addBoth(
+        self,
+        callback: Callable[Concatenate[_T, _P], _T],
+        *args: _P.args,
+        **kwargs: _P.kwargs,
+    ) -> Deferred[_DeferredResultT]:
```

Comment:
Yikes?  This seems like a very unprincipled overload.  I commented it out and mypy didn't give me any errors.  I can't see a reason to _want_ this signature if it isn't being jammed down our throats.

Reply:
This was definitely a mistake. Thank you for catching. I am seriously starting to think about writing a program to emit all these overloads because that might be easier to check for errors than this manual repetition :-|